File: godfather/factions/neutral/arsonist.py
from godfather.game import Game
from godfather.factions.base import Faction

OPPOSING_FACTIONS = ['neutral.serialkiller', 'mafia', 'town']


class ArsonistNeutral(Faction):
    name = 'Arsonist'
    id = 'neutral.arsonist'
    win_con = 'Watch everyone burn.'

    @property
    def category_name(self):
        return 'Neutral'

    def has_won(self, game: Game):
        alive_arsos = len(game.players.filter(
            faction='neutral.arsonist', is_alive=True))
        alive_opposing = len(
            [*filter(lambda pl: pl.role.faction.id in OPPOSING_FACTIONS and pl.is_alive, game.players)])
        return alive_arsos > 0 and alive_opposing == 0

File: src/lib/mafia/factions/neutral/Arsonist.ts
import Faction from '@mafia/structures/Faction';
import type Game from '@mafia/structures/Game';

const WINS_WITH = ['Arsonist', 'Witch', 'Survivor'];

export default class ArsonistFaction extends Faction {
	public name = 'Arsonist';
	public winCondition = 'game/factions:nkWinCondition';

	public hasWon(game: Game) {
		const { players } = game;

		const aliveArsonists = players.filter((player) => player.isAlive && player.role.faction.name === 'Arsonist');
		const aliveOpposing = players.filter((player) => player.isAlive && !WINS_WITH.includes(player.role!.faction.name));

		return aliveArsonists.length > 0 && aliveOpposing.length === 0;
	}
}
